tailcfg/tailcfg.go

```diff
@@ -3030,3 +3034,48 @@ const LBHeader = "Ts-Lb"
 // correspond to those IPs. Any services that don't correspond to a service
 // this client is hosting can be ignored.
 type ServiceIPMappings map[ServiceName][]netip.Addr
+
+// ClientAuditAction represents an auditable action that a client can report to the
+// control plane.  These actions must correspond to the supported actions
+// in the control plane.
+type ClientAuditAction string
+
+const (
+	// AuditNodeDisconnect action should be sent when a node has disconnected
+	// from the control plane.  The details should include a reason in the Details
+	// field, either generated, or entered by the user.
+	AuditNodeDisconnect = ClientAuditAction("DISCONNECT_NODE")
+)
+
+// ClientAuditLog represents a pending audit log request generated by a client.
+//
+// EventID is a random, unique ID for the audit log request.
+// ProfileID is the fixed, unique userProfile ID to whom the audit log is associated.
+// Retries is the number of times we've attempted to submit this log.
+//
+// The Action and Details correspond the the Action, Details and Timestamp fields that will
+// written to the audit log request to the control plane.
+type ClientAuditLog struct {
+	// EventID is a random, unique ID for the audit log request.
+	EventID string `json:",omitempty"`
+	// ProfileID is a the fixed, unique user ipn.ProfileID to whom the audit log is associated.
+	ProfileID string `json:",omitempty"`
+	// Retries is the number of times we've attempted to submit this log.
+	Retries int `json:",omitempty"`
+
+	// Action is the action to be logged. It must correspond to a known action in the control plane.
+	Action ClientAuditAction `json:",omitempty"`
+	// Details is an opaque string, specific to the action being logged
+	Details string `json:",omitempty"`
+	// Timestamp is the time at which the audit log was generated on the node
+	Timestamp time.Time `json:",omitempty"`
+}
+
+// AuditLogRequest represents an audit log request to be sent to the control plane.
+// See ClientAuditLog for a description of the fields
+type AuditLogRequest struct {
+	NodeKey   key.NodePublic    `json:",omitempty"`
+	Action    ClientAuditAction `json:",omitempty"`
+	Details   string            `json:",omitempty"`
+	Timestamp time.Time         `json:",omitempty"`
```

Comment:
Can you please use `omitzero` instead of `omitempty` for `time.Time`?
```suggestion
	Timestamp time.Time         `json:",omitzero"`
```

Reply:
(see https://github.com/golang/go/issues/45669)